File: src/lib/units.ts
// All weights are stored internally as kilograms.
// Display + entry units come from app settings.

export type WeightUnit = 'kg' | 'lb'

const KG_PER_LB = 0.45359237

// Round to 3 decimal places of a kilogram (1 g resolution).
// Avoids long binary-float tails like 1.5875732949999998 from creeping
// into stored values or downstream arithmetic.
export function roundKg(kg: number): number {
  if (!isFinite(kg)) return 0
  return Math.round(kg * 1000) / 1000
}

export function kgToDisplay(kg: number, unit: WeightUnit): number {
  return unit === 'kg' ? kg : kg / KG_PER_LB
}

export function displayToKg(value: number, unit: WeightUnit): number {
  return roundKg(unit === 'kg' ? value : value * KG_PER_LB)
}

export function formatWeight(
  kg: number,
  unit: WeightUnit,
  digits: number = 2,
): string {
  return `${kgToDisplay(kg, unit).toFixed(digits)} ${unit}`
}

export function formatWeightShort(kg: number, unit: WeightUnit): string {
  return kgToDisplay(kg, unit).toFixed(2)
}

File: Services/Units.cs
using RefrigHandle.Models;

namespace RefrigHandle.Services;

public static class Units
{
    public const double KgPerLb = 0.45359237;

    public static double KgToDisplay(double kg, WeightUnit unit)
        => unit == WeightUnit.Kg ? kg : kg / KgPerLb;

    public static double DisplayToKg(double value, WeightUnit unit)
        => unit == WeightUnit.Kg ? value : value * KgPerLb;

    // 1g resolution
    public static double RoundKg(double kg) => Math.Round(kg * 1000) / 1000.0;

    public static string Label(WeightUnit unit) => unit == WeightUnit.Kg ? "kg" : "lb";

    public static string FormatWeight(double kg, WeightUnit unit, int decimals = 2)
    {
        var v = KgToDisplay(kg, unit);
        return v.ToString("F" + decimals) + " " + Label(unit);
    }
}
